File: planner.py
from typing import *
from collections import deque


T = TypeVar("T")


class Graph(Protocol[T]):
    def adjacent(self, node: T) -> Iterable[T]:
        ...


class Path:
    def __init__(self, node: T, steps: int, parent: Optional["Path"]):
        self.node = node
        self.steps = steps
        self.parent = parent

    def as_list(self) -> list[T]:
        return list(self.as_iter())

    def as_iter(self) -> Iterable[T]:
        yield self.node
        if self.parent:
            yield from self.parent.as_iter()

    def __repr__(self):
        return (f"<{self.__class__.__qualname__} "
                f"node={self.node!r} "
                f"steps={self.steps!r} "
                f"parent={(self.parent.node if self.parent else None)!r}"
                ">")


def bft(start: T, graph: Graph[T]) -> Iterable[Path]:
    frontier = deque([Path(node=start, steps=0, parent=None)])
    seen: Set[T] = set()

    while frontier:
        path = frontier.popleft()
        if path.node in seen:
            continue

        yield path
        seen.add(path.node)

        for adj_node in graph.adjacent(path.node):
            frontier.append(Path(
                node=adj_node,
                steps=path.steps + 1,
                parent=path
            ))


def bfs(start: T, goal: T, graph: Graph[T]):
    for path in bft(start, graph):
        if path.node == goal:
            return path

File: src/planner.rs
use std::{
    collections::{HashSet, VecDeque},
    hash::Hash,
};

pub trait Graph<T> {
    type Output: IntoIterator<Item = T>;
    fn adjacent(&self, node: &T) -> Self::Output;
}

#[derive(Clone, Debug)]
pub struct Path<T: Eq + Hash + Clone> {
    node: T,
    steps: usize,
    parent: Box<Option<Path<T>>>,
}

pub fn bft<T: Eq + Hash + Clone>(start: T, graph: impl Graph<T>) -> Vec<Path<T>> {
    let mut frontier = VecDeque::from([Path {
        node: start,
        steps: 0,
        parent: Box::new(Option::None),
    }]);

    let mut seen: HashSet<T> = HashSet::new();

    let mut paths: Vec<Path<T>> = vec![];

    while !frontier.is_empty() {
        let path = frontier.pop_front().unwrap();
        if seen.contains(&path.node) {
            continue;
        }

        paths.push(path.clone());
        seen.insert(path.node.clone());

        for adj_node in graph.adjacent(&path.node) {
            frontier.push_back(Path {
                node: adj_node.clone(),
                steps: path.steps + 1,
                parent: Box::new(Option::Some(path.clone())),
            })
        }
    }

    paths
}

pub fn bfs<T: Eq + Hash + Clone>(start: T, goal: T, graph: impl Graph<T>) -> Option<Path<T>> {
    for path in bft(start, graph) {
        if path.node == goal {
            return Option::Some(path);
        }
    }
    Option::None
}
